src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,212 @@
+| `JSPromise<T extends JSAny?>`       | `Future<T extends JSAny?>`               |
+| Typed arrays e.g. `JSUint8Array`    | `dart:typed_data`                        |
+| `JSBoxedDartObject`                 | Opaque Dart value                        |
+
+**Important**: Conversions may have different costs depending on the compiler.
+Prefer to only convert when you need to. Furthermore, conversions may or may not
+produce a new value. This doesn’t matter for immutable values like numbers, but
+does matter for types like `List`s. A conversion to a `JSArray` may produce a
+new value by copying or may not, so do not rely on modifications to the
+`JSArray` to affect the `List`. Typed array conversions have a similar
+limitation. Look at the specific conversion function for more details.
+
+## Requirements on `external` declarations and `Function.toJS`
+
+In order to ensure type safety and consistency, the compiler places requirements
+on what types can flow into and out of JS. Passing arbitrary Dart values into JS
+is not allowed. Instead, the compiler requires users to use a compatible interop
+type like a JS type or a primitive, which would then be implicitly converted by
+the compiler. For example, these would be allowed:
+
+{:.good}
+```dart
+@JS()
+external void primitives(String a, int b, double c, num d, bool e);
+```
+
+{:.good}
+```dart
+@JS()
+external JSArray jsTypes(JSObject _, JSString __);
+```
+
+{:.good}
+```dart
+extension type InteropType(JSObject _) {}
+
+@JS()
+external InteropType get interopType;
+```
+
+whereas these would return an error:
+
+{:.bad}
+```dart
+@JS()
+external Function get function;
+```
+
+{:.bad}
+```dart
+@JS()
+external set list(List _);
+```
+
+These same requirements exist when you use [`Function.toJS`] to make a Dart
+function callable in JS. The values that flow into and out of this callback must
+be a compatible interop type or a primitive.
+
+If you use a Dart primitive e.g. `String`, an implicit conversion happens in the
+compiler to convert that value from a JS value to a Dart value. If performance
+is critical and you don’t need to examine the contents of the string, then using
+`JSString` instead to avoid the conversion cost may make sense.
+
+## Compatibility, type checks, and casts
+
+As mentioned above, the representation type of JS types may differ based on the
+compiler. This affects runtime type-checking and casts. Therefore, almost always
+avoid `is` checks where the value is an interop type or where the target type is
+an interop type. Avoid casts between Dart types and interop types. In order to
+type-check a JS value, use an interop member like [`typeofEquals`] or
+[`instanceOfString`] that examines the JS value itself.
+
+Bad:
+
+{:.bad}
+```dart
+void f(JSAny a) {
+  if (a is String) { … }
+}
+```
+
+{:.bad}
+```dart
+void f(JSObject o) {
+  if (o is JSObject) { … }
+}
+```
+
+{:.bad}
+```dart
+void f(JSString s) {
+  s as String;
+}
+```
+
+Okay:
```

Comment:
Removed "As mentioned..." since people might end up here following a link directly to this section (and the above mention is pretty light anyway)

I didn't rewrite anything else, just moved the sentences describing each situation to right about the code block showing that situation (so you don't have to say good/bad)

````suggestion
The representation type of JS types may differ based on the compiler.
This affects runtime type-checking and casts. Therefore, almost always
avoid `is` checks where the value is an interop type or where the target type is
an interop type:

{:.bad}
```dart
void f(JSAny a) {
  if (a is String) { … }
}
```

{:.bad}
```dart
void f(JSObject o) {
  if (o is JSObject) { … }
}
```

Also, avoid casts between Dart types and interop types:

{:.bad}
```dart
void f(JSString s) {
  s as String;
}
```

To type-check a JS value, use an interop member like [`typeofEquals`] or
[`instanceOfString`] that examines the JS value itself:
````

Reply:
I actually just landed a better way for users to do this, so I'll add a TODO to link that instead once it's in a dev release. :D Users won't have access to it until 3.4, so I'll still mention the other two and add examples.